pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator.java

```diff
@@ -0,0 +1,193 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.modeldefinition.generator;
+
+import hudson.Extension;
+import hudson.Functions;
+import hudson.model.Action;
+import hudson.model.Descriptor;
+import hudson.model.Item;
+import hudson.model.Job;
+import jenkins.model.Jenkins;
+import jenkins.model.TransientActionFactory;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.structs.SymbolLookup;
+import org.jenkinsci.plugins.structs.describable.UninstantiatedDescribable;
+import org.jenkinsci.plugins.workflow.cps.Snippetizer;
+import org.jenkinsci.plugins.workflow.cps.SnippetizerLink;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.HttpResponses;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension
+public class DirectiveGenerator extends Snippetizer {
+    private static final Logger LOGGER = Logger.getLogger(DirectiveGenerator.class.getName());
+
+    public static final String ACTION_URL = "directive-generator";
+
+    @Restricted(NoExternalUse.class)
+    public static final String GENERATE_URL = ACTION_URL + "/generateDirective";
+
+    @Override public String getUrlName() {
+        return ACTION_URL;
+    }
+
+    @Nonnull
+    public List<DirectiveDescriptor> getDirectives() {
+        return DirectiveDescriptor.all();
+    }
+
+    @Restricted(DoNotUse.class) // accessed via REST API
+    public HttpResponse doGenerateDirective(StaplerRequest req, @QueryParameter String json) throws Exception {
+        // TODO is there not an easier way to do this? Maybe Descriptor.newInstancesFromHeteroList on a one-element JSONArray?
+        JSONObject jsonO = JSONObject.fromObject(json);
+        Jenkins j = Jenkins.getActiveInstance();
+        Class<?> c = j.getPluginManager().uberClassLoader.loadClass(jsonO.getString("stapler-class"));
+        DirectiveDescriptor descriptor = (DirectiveDescriptor)j.getDescriptor(c.asSubclass(AbstractDirective.class));
+        if (descriptor == null) {
+            return HttpResponses.plainText("<could not find " + c.getName() + ">");
```

Comment:
Shouldn't this be more of a 404 or something?

Reply:
This is trying to fit the behavior we already have with Snippet Generator - we just need to get a string to show in the result box, and a 404 would break things.